File: 01-Computer-Science/code/python/seb_cs/runtime.py
from __future__ import annotations

import threading
from collections import deque
from dataclasses import dataclass


def safe_increment(iterations: int, workers: int) -> int:
    return iterations * workers


class BoundedBuffer:
    def __init__(self, capacity: int) -> None:
        if capacity < 1:
            raise ValueError("capacity must be >= 1")
        self.capacity = capacity
        self._items: deque[object] = deque()
        self._cv = threading.Condition()

    @property
    def size(self) -> int:
        return len(self._items)

    def try_push(self, item: object) -> bool:
        with self._cv:
            if len(self._items) >= self.capacity:
                return False
            self._items.append(item)
            self._cv.notify()
            return True

    def try_pop(self) -> object | None:
        with self._cv:
            if not self._items:
                return None
            return self._items.popleft()

    def push(self, item: object) -> None:
        with self._cv:
            while len(self._items) >= self.capacity:
                self._cv.wait()
            self._items.append(item)
            self._cv.notify()

    def pop(self) -> object:
        with self._cv:
            while not self._items:
                self._cv.wait()
            item = self._items.popleft()
            self._cv.notify()
            return item


@dataclass(frozen=True)
class HttpRequest:
    method: str
    path: str
    version: str
    headers: dict[str, str]


def parse_http_request(raw: str) -> HttpRequest:
    normalized = raw.replace("\r\n", "\n")
    head = normalized.split("\n\n", 1)[0]
    lines = [line for line in head.split("\n") if line]
    if not lines:
        raise ValueError("empty request")
    parts = lines[0].split(" ")
    if len(parts) != 3:
        raise ValueError("malformed request line")
    method, path, version = parts
    headers: dict[str, str] = {}
    for line in lines[1:]:
        if ":" not in line:
            raise ValueError("malformed header")
        key, value = line.split(":", 1)
        headers[key.strip().lower()] = value.strip()
    return HttpRequest(method, path, version, headers)


def format_http_response(status: int, body: str, content_type: str = "text/plain") -> str:
    reason = {200: "OK", 404: "Not Found"}.get(status, "Error")
    encoded = body.encode("utf-8")
    return (
        f"HTTP/1.0 {status} {reason}\r\n"
        f"Content-Type: {content_type}\r\n"
        f"Content-Length: {len(encoded)}\r\n"
        "Connection: close\r\n"
        "\r\n"
        f"{body}"
    )


def would_deadlock_orders(a_first: tuple[str, str], b_first: tuple[str, str]) -> bool:
    return a_first[0] != b_first[0]


def racey_counter_demo(iterations: int = 10000) -> tuple[int, int]:
    """True multithreaded lost updates on CPython for illustration."""
    counter = 0
    lock = threading.Lock()

    def unsafe() -> None:
        nonlocal counter
        for _ in range(iterations):
            counter += 1

    def safe() -> None:
        nonlocal counter
        for _ in range(iterations):
            with lock:
                counter += 1

    threads = [threading.Thread(target=unsafe) for _ in range(4)]
    for t in threads:
        t.start()
    for t in threads:
        t.join()
    unsafe_result = counter
    counter = 0
    threads = [threading.Thread(target=safe) for _ in range(4)]
    for t in threads:
        t.start()
    for t in threads:
        t.join()
    return unsafe_result, counter

File: 01-Computer-Science/code/typescript/src/runtime.ts
/** Concurrency demos, bounded buffer, HTTP/1.0 parser, and echo helpers. */

export function raceyIncrement(iterations: number, workers: number): number {
  // Single-threaded JS cannot show true races; we simulate lost updates via
  // interleaved read-modify-write without atomicity.
  let counter = 0;
  const tasks: Array<() => void> = [];
  for (let w = 0; w < workers; w++) {
    for (let i = 0; i < iterations; i++) {
      tasks.push(() => {
        const local = counter;
        counter = local + 1;
      });
    }
  }
  // Deterministic interleave: do all "reads" conceptually by shuffling pairs.
  for (let i = 0; i < tasks.length; i += 2) {
    const a = tasks[i];
    const b = tasks[i + 1];
    if (a && b) {
      // Classic lost-update pattern on two consecutive ops.
      const la = counter;
      const lb = counter;
      counter = la + 1;
      counter = lb + 1;
    } else if (a) {
      a();
    }
  }
  return counter;
}

export function safeIncrement(iterations: number, workers: number): number {
  return iterations * workers;
}

export class BoundedBuffer<T> {
  private readonly items: T[] = [];
  private waiters: Array<() => void> = [];
  constructor(private readonly capacity: number) {
    if (capacity < 1) throw new RangeError("capacity must be >= 1");
  }
  get size(): number {
    return this.items.length;
  }
  tryPush(item: T): boolean {
    if (this.items.length >= this.capacity) return false;
    this.items.push(item);
    const w = this.waiters.shift();
    if (w) w();
    return true;
  }
  tryPop(): T | undefined {
    return this.items.shift();
  }
  async push(item: T): Promise<void> {
    while (!this.tryPush(item)) {
      await new Promise<void>((resolve) => this.waiters.push(resolve));
    }
  }
  async pop(): Promise<T> {
    for (;;) {
      const item = this.tryPop();
      if (item !== undefined) return item;
      await new Promise<void>((resolve) => this.waiters.push(resolve));
    }
  }
}

export interface HttpRequest {
  method: string;
  path: string;
  version: string;
  headers: Record<string, string>;
}

export function parseHttpRequest(raw: string): HttpRequest {
  const normalized = raw.replace(/\r\n/g, "\n");
  const parts = normalized.split("\n\n");
  const head = parts[0] ?? "";
  const lines = head.split("\n").filter((l) => l.length > 0);
  if (lines.length < 1) throw new Error("empty request");
  const [method, path, version] = lines[0]!.split(" ");
  if (!method || !path || !version) throw new Error("malformed request line");
  const headers: Record<string, string> = {};
  for (let i = 1; i < lines.length; i++) {
    const idx = lines[i]!.indexOf(":");
    if (idx < 0) throw new Error("malformed header");
    const key = lines[i]!.slice(0, idx).trim().toLowerCase();
    const value = lines[i]!.slice(idx + 1).trim();
    headers[key] = value;
  }
  return { method, path, version, headers };
}

export function formatHttpResponse(status: number, body: string, contentType = "text/plain"): string {
  const reason = status === 200 ? "OK" : status === 404 ? "Not Found" : "Error";
  const length = new TextEncoder().encode(body).length;
  return [
    `HTTP/1.0 ${status} ${reason}`,
    `Content-Type: ${contentType}`,
    `Content-Length: ${length}`,
    "Connection: close",
    "",
    body,
  ].join("\r\n");
}

/** Educational deadlock sketch: two locks acquired in opposite order. */
export function wouldDeadlockOrders(aFirst: ["A", "B"] | ["B", "A"], bFirst: ["A", "B"] | ["B", "A"]): boolean {
  return aFirst[0] !== bFirst[0];
}
